RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,377 @@
+   <td>Total
+   </td>
+   <td>138
+   </td>
+  </tr>
+</table>
+
+
+We can maintain a mapping of metric key (which can be anywhere between 30 to 50 characters) and metric value in the implementation of BaseStatsReporter class. Each of the above metric values is stored in a 8 byte integer type. At any moment we may have 138 × 8 = 1104 bytes of values data held in-memory. Also, we have 138 × ~50 = 6900 ~ 7K bytes for keys. In total we have 7K + 1104 ~ 8K bytes of memory. Thus holding metrics in memory has the least impact on Prestissimo workers.
+
+Given the above estimates, we decide to keep the stats in-memory and we can revisit this if the memory footprint grows.
+
+Note: Estimating the size of a histogram with quantiles is not straight forward, it depends on rotation time and error tolerance settings.
+
+
+### 
+**Metric timestamps**
+
+Since we overwrite metrics in our current design, it is likely that Prometheus may miss updates as it pulls metrics at regular intervals. The timestamps are assigned by the Prometheus server (this is recommended) and may not reflect the exact timestamp at which the metric was recorded. On the other hand, we can maintain a list of metric values and timestamps seen so far in Prestissimo and in the response to the Prometheus we can include these &lt;timestamp, value> pairs. Prometheus must be configured to honour these timestamps. But this approach is not recommended.
+
+Prestissimo emits metric values without timestamps. Prometheus DB will be assigning timestamps to the metric values after it pulls from the Prestissimo endpoint.
+
+
+### 
+**Serialisation Format**
+
+
+#### 
+**Serialize using custom interface**
+
+By default, we shall implement JSON and Prometheus Data Model serialisation of metrics. We shall expose a new Serialization interface that users can implement to customize serialization. Pros: In-house and no external dependencies. Cons: It could be challenging to keep it in sync with Prometheus data model versions. Adding support for histogram quantiles is not simple.
+
+
+#### 
+**Serializing using Prometheus-CPP: (Preferred)**
+
+Pros: Popular and simplifies histogram and summary metric maintenance. The library has implemented unit tests and integration tests for serialisation and data correctness. Cons: External dependency.
+
+Both of these approaches are prototyped in [this PR](https://github.com/prestodb/presto/pull/21599/files#).
+
+
+## 
+**Configuration**
+
+Runtime configs:
+
+1. Currently, we have `runtime-metrics-collection-enabled` configuration property in Native Presto, which when set to true, starts recording metrics.
+2. (TBD) Configs to optionally enable histogram/summary type metrics which may not be useful for day to day monitoring.
+3. (TBD) Mechanism to white list metrics. Compile time configs: Also, proposing a compile to time config PRESTO_ENABLE_PROMETHEUS, which when turned ON, includes prometheus-metrics directory and registers PrometheusReporter as the metrics reporter.
```

Comment:
Do we need 2 if we implement 3?
Probably move 3 to a future work section?

Reply:
Done.